core/src/main/scala/caliban/schema/ArgBuilder.scala

```diff
@@ -205,6 +209,19 @@ trait ArgBuilderInstances extends ArgBuilderDerivation {
 
   implicit lazy val upload: ArgBuilder[Upload] = {
     case StringValue(v) => Right(Upload(v))
-    case other          => Left(ExecutionError(s"Can't build an Upload from $other"))
+    case other          => Left(MkInvalidInputArgument("Upload", other.toString))
+  }
+}
+
+case object InvalidInputArgument extends NoStackTrace {
+  override def getMessage: String = "invalid input argument"
+}
+
+private object MkInvalidInputArgument {
+  private val vowels = BitSet(List('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U').map(_.toInt): _*)
+
+  def apply(expected: String, actual: String): ExecutionError = {
+    val prefix = if (expected.nonEmpty && vowels.contains(expected.codePointAt(0))) "an" else "a"
+    ExecutionError(s"Can't build $prefix $expected from $actual", innerThrowable = Some(InvalidInputArgument))
```

Comment:
I'd just cheat and change to "build an instance of" 😅

Reply:
Yeah good idea. I was thinking of preserving the error messages as much as possible but to be honest I don't think that's necessary